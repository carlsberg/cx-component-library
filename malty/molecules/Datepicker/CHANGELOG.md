# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.2.0 (2024-05-31)

### Bug Fixes

- **datepicker:** [DSM-547] Datepicker unwanted text ([#560](https://github.com/CarlsbergGBS/cx-component-library/issues/560)) ([3e3e31e](https://github.com/CarlsbergGBS/cx-component-library/commit/3e3e31e31b45c51f6e6156dfcadf1bac1668a095))
- **datepicker:** [DSM-572] Add readonly on input when selectsRange ([#617](https://github.com/CarlsbergGBS/cx-component-library/issues/617)) ([58e47fc](https://github.com/CarlsbergGBS/cx-component-library/commit/58e47fc3c33d97e92d30d119b4a65063d019aa6b))
- **datepicker:** [DSM-824] add dateFormat prop to the component ([#833](https://github.com/CarlsbergGBS/cx-component-library/issues/833)) ([1240318](https://github.com/CarlsbergGBS/cx-component-library/commit/12403188db910f018c2c669f5911b007dcc808dd))
- **DSM-235:** fixed cursor when readOnly ([#201](https://github.com/CarlsbergGBS/cx-component-library/issues/201)) ([b3798ec](https://github.com/CarlsbergGBS/cx-component-library/commit/b3798ecb04dd1806f85fd877f271661457f26825))
- **DSM-312:** fixed rounded corners on mobile ([#250](https://github.com/CarlsbergGBS/cx-component-library/issues/250)) ([8eff302](https://github.com/CarlsbergGBS/cx-component-library/commit/8eff302cde8ad59e993cf082ddc8f68e383e9e54))
- **DSM-315:** fix rounded corners ([#272](https://github.com/CarlsbergGBS/cx-component-library/issues/272)) ([396a173](https://github.com/CarlsbergGBS/cx-component-library/commit/396a1732011c8936140ecee20070612230ec2444))
- **DSM-387:** Added missing onClick handler in buttons and opening and closing logic ([#315](https://github.com/CarlsbergGBS/cx-component-library/issues/315)) ([e9e8aed](https://github.com/CarlsbergGBS/cx-component-library/commit/e9e8aed6f0320b90bd58ba67aebd42d0fef04456))
- **DSM-411:** fixed datepicker minor bugs ([#340](https://github.com/CarlsbergGBS/cx-component-library/issues/340)) ([5303712](https://github.com/CarlsbergGBS/cx-component-library/commit/53037129cc406b8ae90f557b28b6af8c75739975))
- Minor component fixes ([#265](https://github.com/CarlsbergGBS/cx-component-library/issues/265)) ([c75bd7c](https://github.com/CarlsbergGBS/cx-component-library/commit/c75bd7c926163db6efb65ab89a64d8637d7a4664))
- **pagination:** [DSM-589] Refactor pagination and add validations ([#629](https://github.com/CarlsbergGBS/cx-component-library/issues/629)) ([caa545d](https://github.com/CarlsbergGBS/cx-component-library/commit/caa545dad76254fcc0893dc4c76cf0792ae40751))
- Remove typography provider from every component ([#383](https://github.com/CarlsbergGBS/cx-component-library/issues/383)) ([5d05680](https://github.com/CarlsbergGBS/cx-component-library/commit/5d05680e40d3da5bb3042bd8191427a70606afc3))
- reverted changes ([#396](https://github.com/CarlsbergGBS/cx-component-library/issues/396)) ([ecfcf57](https://github.com/CarlsbergGBS/cx-component-library/commit/ecfcf573863253fea16e24764b6f452c3cc4f215))
- Run auto fixable lint based on new `eslint` configuration ([#501](https://github.com/CarlsbergGBS/cx-component-library/issues/501)) ([9ee94cc](https://github.com/CarlsbergGBS/cx-component-library/commit/9ee94cc2cd407f717e62d8857b5f20a74e7bbec4))

### Features

- add tsconfig files and update package.json files for each component ([3bfc644](https://github.com/CarlsbergGBS/cx-component-library/commit/3bfc644e1cfc9dbb7cf7a0469e25fce055b53240))
- additional stories ([#291](https://github.com/CarlsbergGBS/cx-component-library/issues/291)) ([438af3b](https://github.com/CarlsbergGBS/cx-component-library/commit/438af3b969e0f5b4b908436ee58b93c4b9ddc97c))
- change organization name ([a0aa428](https://github.com/CarlsbergGBS/cx-component-library/commit/a0aa428f43138af5707a3ff4c0e36b7c056f02e6))
- **datepicker:** [DSM-799] add tests ([#783](https://github.com/CarlsbergGBS/cx-component-library/issues/783)) ([84ab41d](https://github.com/CarlsbergGBS/cx-component-library/commit/84ab41d714bf280d29454c33c9f50e6bba29ed63))
- **datepicker:** [DSM-967] add support for `onClose` callback ([#853](https://github.com/CarlsbergGBS/cx-component-library/issues/853)) ([6226f88](https://github.com/CarlsbergGBS/cx-component-library/commit/6226f8873cc77e92bef258b12c86562ad06de3ae))
- **DSM-119:** Datepicker background bug fix ([#164](https://github.com/CarlsbergGBS/cx-component-library/issues/164)) ([bc14919](https://github.com/CarlsbergGBS/cx-component-library/commit/bc14919e487ec9051d78e535adb0a0653385c0fc))
- **DSM-228:** Added rating component ([#379](https://github.com/CarlsbergGBS/cx-component-library/issues/379)) ([ab47354](https://github.com/CarlsbergGBS/cx-component-library/commit/ab47354db803fb051a1f87d66ee34966698df8bd))
- **DSM-235:** added readOnly on inputs ([#181](https://github.com/CarlsbergGBS/cx-component-library/issues/181)) ([36c5ef7](https://github.com/CarlsbergGBS/cx-component-library/commit/36c5ef7cb1225bf5a150918e01e10fd253377110))
- **DSM-248:** Headline HTML tags ([#184](https://github.com/CarlsbergGBS/cx-component-library/issues/184)) ([305f5e0](https://github.com/CarlsbergGBS/cx-component-library/commit/305f5e0fa92d5813f4877b338ad827a8c347d452))
- **DSM-362:** snapshots removed & unit test updates ([#307](https://github.com/CarlsbergGBS/cx-component-library/issues/307)) ([78423eb](https://github.com/CarlsbergGBS/cx-component-library/commit/78423eb5cfe04fe8a3a7af1f5af074ac6d35e235))
- **DSM-368:** Required prop & label component ([#384](https://github.com/CarlsbergGBS/cx-component-library/issues/384)) ([983bd78](https://github.com/CarlsbergGBS/cx-component-library/commit/983bd78c8f526a8faa2df601bf918b40219820f2))
- **DSM-385:** added dropdown to select year ([#368](https://github.com/CarlsbergGBS/cx-component-library/issues/368)) ([fe53588](https://github.com/CarlsbergGBS/cx-component-library/commit/fe53588f2142217b516aad4b71b0f934b391208e))
- **DSM-387:** datepicker include actions and captions ([#308](https://github.com/CarlsbergGBS/cx-component-library/issues/308)) ([eafd4ba](https://github.com/CarlsbergGBS/cx-component-library/commit/eafd4ba9a4b1a28d73dc0c569051548445aaad1a))
- **DSM-439:** Change font-family behavior ([#470](https://github.com/CarlsbergGBS/cx-component-library/issues/470)) ([ab6dad4](https://github.com/CarlsbergGBS/cx-component-library/commit/ab6dad414c42c4343de3de6abaa4cd3ff98cfca1))
- **DSM-513:** Add new behaviors in Datepicker ([#512](https://github.com/CarlsbergGBS/cx-component-library/issues/512)) ([098378c](https://github.com/CarlsbergGBS/cx-component-library/commit/098378c23b8e5d42359d26a3896b7a7f888e8d43))
- **general:** [DSM-953] refactor icons implementation ([#892](https://github.com/CarlsbergGBS/cx-component-library/issues/892)) ([08d259f](https://github.com/CarlsbergGBS/cx-component-library/commit/08d259f476ce97e9db8ac41d5eb4773eabca0d37))
- **global:** [DSM-784] add space props in all components ([#842](https://github.com/CarlsbergGBS/cx-component-library/issues/842)) ([9f03f59](https://github.com/CarlsbergGBS/cx-component-library/commit/9f03f59e7518a99abcea06bacbee989d1ddb0465))
- **global:** migrate from bit to lerna ([a0110c8](https://github.com/CarlsbergGBS/cx-component-library/commit/a0110c8831370dc762c193b17cc593eed381f990))
- **global:** Remove `react/jsx-props-no-spreading` comments ([#558](https://github.com/CarlsbergGBS/cx-component-library/issues/558)) ([823444c](https://github.com/CarlsbergGBS/cx-component-library/commit/823444ce0800bc7f49b584de59d247eb8497f559))
- **global:** update storybook files to be compliant with v7 ([#970](https://github.com/CarlsbergGBS/cx-component-library/issues/970)) ([c61d4b4](https://github.com/CarlsbergGBS/cx-component-library/commit/c61d4b40a0755becf942ad3f28758a159f8c54e4))
- new icon colors ([#215](https://github.com/CarlsbergGBS/cx-component-library/issues/215)) ([9247d33](https://github.com/CarlsbergGBS/cx-component-library/commit/9247d3370bcde78432033ba8b23837f97e0b0dc0))
- remove empty composition files ([f8173f4](https://github.com/CarlsbergGBS/cx-component-library/commit/f8173f4a2ecbf80bb7b6ffe848c023ae31819c2d))
- Several component updates ([#185](https://github.com/CarlsbergGBS/cx-component-library/issues/185)) ([633c9f0](https://github.com/CarlsbergGBS/cx-component-library/commit/633c9f0a4432e9e3d08ff05156453660929c19a2))
- Use new `eslint-config-carlsberg` with streamlined way to create components ([#559](https://github.com/CarlsbergGBS/cx-component-library/issues/559)) ([a459332](https://github.com/CarlsbergGBS/cx-component-library/commit/a45933215bc0b523220743f17d988d7f1ad5fce6))
- v1.0.22 ([#194](https://github.com/CarlsbergGBS/cx-component-library/issues/194)) ([885949f](https://github.com/CarlsbergGBS/cx-component-library/commit/885949fdf10b8883c0f4f45447fc5e5711bc5b3f))
