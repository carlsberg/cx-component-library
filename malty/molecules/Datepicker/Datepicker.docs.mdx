---
labels: ['react', 'typescript', 'datepicker']
description: 'A Datepicker component.'
---
