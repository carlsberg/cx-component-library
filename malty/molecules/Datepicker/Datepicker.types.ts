export interface DatepickerProps {
  label: string;
  startDate: Date | null;
  onChange: (date: (Date | null) | [Date | null, Date | null]) => void;
  locale?: string;
  minDate?: Date;
  maxDate?: Date;
  disabled?: boolean;
  excludeDates?: Date[];
  placeholderText?: string;
  selectsRange?: boolean;
  endDate?: Date | null;
  inline?: boolean;
  dateFormat?: string;
  readOnly?: boolean;
  captions?: Caption[];
  primaryAction?: Action;
  secondaryAction?: Action;
}

export type Caption = {
  color: string;
  copy: string;
  borderColor?: string;
};

export type Action = {
  copy: string;
  action?: () => void;
};
