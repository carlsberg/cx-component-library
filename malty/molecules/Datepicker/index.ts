export { Datepicker } from './Datepicker';
export { Colors } from './Datepicker.types';
export type { DatepickerProps } from './Datepicker.types';
