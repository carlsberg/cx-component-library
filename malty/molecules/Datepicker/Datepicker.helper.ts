import { globalTheme as defaultTheme } from '@carlsberggroup/malty.theme.malty-theme-provider';
import { useContext, useMemo } from 'react';
import { ThemeContext } from 'styled-components';
import { Colors } from './Datepicker.types';

export function useColorsMapping() {
  const theme = useContext(ThemeContext) || defaultTheme;
  const mappedColors = useMemo(
    () => ({
      [Colors.DigitalBlack]: theme.colors.colours.default['digital-black'].value,
      [Colors.Transparent]: theme.colors.colours.default.transparent.value,
      [Colors.White]: theme.colors.colours.default.white.value,
      [Colors.InformationArchive]: theme.colors.colours.information.archive.value,
      [Colors.InformationClose]: theme.colors.colours.information.close.value,
      [Colors.InformationDisable]: theme.colors.colours.information.disable.value,
      [Colors.InformationHold]: theme.colors.colours.information.hold.value,
      [Colors.InformationIndirect]: theme.colors.colours.information.indirect.value,
      [Colors.InformationLive]: theme.colors.colours.information.live.value,
      [Colors.InformationMultiple]: theme.colors.colours.information.multiple.value,
      [Colors.InformationNew]: theme.colors.colours.information.new.value,
      [Colors.InformationParked]: theme.colors.colours.information.parked.value,
      [Colors.InformationProspect]: theme.colors.colours.information.prospect.value,
      [Colors.InformationUpdate]: theme.colors.colours.information.update.value,
      [Colors.InformationWholesaler]: theme.colors.colours.information.wholesaler.value,
      [Colors.OverlayDigitalBlack5]: theme.colors.colours.overlay['digital-black'][5].value,
      [Colors.OverlayDigitalBlack10]: theme.colors.colours.overlay['digital-black'][10].value,
      [Colors.OverlayDigitalBlack25]: theme.colors.colours.overlay['digital-black'][25].value,
      [Colors.OverlayDigitalBlack50]: theme.colors.colours.overlay['digital-black'][50].value,
      [Colors.OverlayDigitalBlack75]: theme.colors.colours.overlay['digital-black'][75].value,
      [Colors.OverlayDigitalBlack90]: theme.colors.colours.overlay['digital-black'][90].value,
      [Colors.OverlayWhite5]: theme.colors.colours.overlay.white[5].value,
      [Colors.OverlayWhite10]: theme.colors.colours.overlay.white[10].value,
      [Colors.OverlayWhite25]: theme.colors.colours.overlay.white[25].value,
      [Colors.OverlayWhite50]: theme.colors.colours.overlay.white[50].value,
      [Colors.OverlayWhite75]: theme.colors.colours.overlay.white[75].value,
      [Colors.OverlayWhite90]: theme.colors.colours.overlay.white[90].value,
      [Colors.Support20]: theme.colors.colours.support[20].value,
      [Colors.Support40]: theme.colors.colours.support[40].value,
      [Colors.Support60]: theme.colors.colours.support[60].value,
      [Colors.Support80]: theme.colors.colours.support[80].value,
      [Colors.Support100]: theme.colors.colours.support[100].value,
      [Colors.SystemAlertLight]: theme.colors.colours.system['alert-light'].value,
      [Colors.SystemAlertStrong]: theme.colors.colours.system['alert-strong'].value,
      [Colors.SystemDisableLightTheme]: theme.colors.colours.system['disable-light-theme'].value,
      [Colors.SystemDisableDarkTheme]: theme.colors.colours.system['disable-dark-theme'].value,
      [Colors.SystemFail]: theme.colors.colours.system.fail.value,
      [Colors.SystemFailLight]: theme.colors.colours.system['fail-light'].value,
      [Colors.SystemNotificationLight]: theme.colors.colours.system['notification-light'].value,
      [Colors.SystemNotificationStrong]: theme.colors.colours.system['notification-strong'].value,
      [Colors.SystemSuccess]: theme.colors.colours.system.success.value,
      [Colors.SystemSuccessLight]: theme.colors.colours.system['success-light'].value
    }),
    [theme]
  );

  return mappedColors;
}
