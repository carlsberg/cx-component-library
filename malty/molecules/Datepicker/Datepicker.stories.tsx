import { generateStorybookSpacing } from '@carlsberggroup/malty.utils.space';
import { Meta, StoryObj } from '@storybook/react';
import React, { useEffect, useState } from 'react';
import { Datepicker, DATE_PICKER_DEFAULT_FORMAT } from './Datepicker';
import { Colors, DatepickerProps, DatepickerSize } from './Datepicker.types';

const DatepickerComponent = (props: DatepickerProps) => {
  const { selectsRange, dateFormat, startDate, endDate } = props;
  const [newStartDate, setNewStartDate] = useState<Date | null>(startDate);

  useEffect(() => {
    setNewStartDate(startDate);
  }, [startDate]);

  const [newEndDate, setNewEndDate] = useState<Date | null | undefined>(endDate);

  useEffect(() => {
    setNewEndDate(endDate);
  }, [endDate]);

  const onChange = (date: (Date | null) | [Date | null, Date | null]) => {
    let start;
    let end;

    if (Array.isArray(date)) {
      [start, end] = date;
      setNewStartDate(start);
      setNewEndDate(end);
    }
  };

  const onChangeStartDate = (date: (Date | null) | [Date | null, Date | null]) => {
    if (!Array.isArray(date)) {
      setNewStartDate(date);
    }
  };

  return (
    <div style={{ height: '560px' }}>
      <Datepicker
        {...props}
        dateFormat={dateFormat === '' ? DATE_PICKER_DEFAULT_FORMAT : dateFormat}
        startDate={newStartDate}
        endDate={newEndDate}
        onChange={selectsRange ? onChange : onChangeStartDate}
      />
    </div>
  );
};

const meta: Meta<DatepickerProps> = {
  component: Datepicker,
  title: 'Forms/Datepicker',
  parameters: {
    importObject: 'Datepicker',
    importPath: '@carlsberggroup/malty.molecules.datepicker'
  },
  render: (args) => <DatepickerComponent {...args} />,
  argTypes: {
    label: {
      description: 'The datepicker input label',
      control: 'text'
    },
    startDate: {
      description: 'Initial selected date',
      control: 'date'
    },
    endDate: {
      description: 'Final selected date. Is only visible if the selectRanges prop is set to true',
      control: 'date',
      if: { arg: 'selectsRange' }
    },
    onChange: {
      description: 'Action to perform when clicking a calendar day'
    },
    onClose: {
      description: 'Action to perform when datepicker is closed'
    },
    minDate: {
      description: 'Disable days before defined min. date',
      control: 'date'
    },
    maxDate: {
      description: 'Disable days after defined max. date',
      control: 'date'
    },
    dateFormat: {
      description: `Custom date format, default is ${DATE_PICKER_DEFAULT_FORMAT}`,
      control: 'text'
    },
    excludeDates: {
      description: 'Disable array of days',
      control: 'array'
    },
    size: {
      description: 'Button size. Options are',
      options: Object.values(DatepickerSize),
      control: 'select'
    },
    disabled: {
      description: 'Disable datepicker',
      control: 'boolean'
    },
    readOnly: {
      control: 'boolean',
      description: 'ReadOnly datepicker'
    },
    selectsRange: {
      description: 'Enable date range selection. This will set the input as readonly',
      control: 'boolean'
    },
    inline: {
      description: 'Display calendar without input',
      control: 'boolean'
    },
    placeholderText: {
      description: 'Input placeholder',
      control: 'text'
    },
    locale: {
      description: 'ISO language code',
      control: 'text'
    },
    captions: {
      description: 'Captions for datepicker',
      control: 'array'
    },
    primaryAction: {
      description: 'Apply date',
      control: 'object'
    },
    secondaryAction: {
      description: 'Cancel apply date',
      control: 'object'
    },
    shouldCloseOnSelect: {
      description: 'Whether the datepicker should close automatically upon selection',
      control: 'boolean'
    },
    withPortal: {
      description: 'Whether the datepicker should open within a [Portal](https://reactjs.org/docs/portals.html)',
      control: 'boolean'
    },
    required: {
      control: 'boolean',
      description: 'Makes the Datepicker required to fill'
    },
    withoutBorder: {
      control: 'boolean',
      description: 'Disables the border and padding from the calendar portal only if inline is true'
    },
    ...generateStorybookSpacing()
  }
};

type Story = StoryObj<DatepickerProps>;

export const Base: Story = {
  args: {
    label: 'Select date',
    required: false,
    startDate: new Date(),
    dateFormat: DATE_PICKER_DEFAULT_FORMAT
  }
};

export const Disabled: Story = {
  args: {
    ...Base.args,
    disabled: true
  }
};

export const Readonly: Story = {
  args: {
    ...Base.args,
    readOnly: true
  }
};

export const Range: Story = {
  args: {
    ...Base.args,
    endDate: new Date(),
    selectsRange: true
  }
};

export const CaptionsAndActions: Story = {
  args: {
    ...Base.args,
    captions: [
      {
        label: 'Selected',
        color: Colors.DigitalBlack
      },
      {
        label: 'Today',
        color: Colors.SystemFail
      },
      {
        label: 'Available if you order until 5pm',
        color: Colors.White,
        borderColor: Colors.InformationIndirect,
        dotted: true
      },
      {
        label: 'Order placed',
        color: Colors.SystemSuccess
      }
    ],
    primaryAction: {
      label: 'Apply',
      action: () => true
    },
    secondaryAction: {
      label: 'Cancel',
      action: () => true
    }
  }
};

export default meta;
