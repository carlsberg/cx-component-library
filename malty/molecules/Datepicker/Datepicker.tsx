import { IconColor, IconSize } from '@carlsberggroup/malty.atoms.icon-wrapper';
import Calendar from '@carlsberggroup/malty.icons.calendar';
import { globalTheme as defaultTheme, TypographyProvider } from '@carlsberggroup/malty.theme.malty-theme-provider';
import React, { ReactNode, useContext } from 'react';
import DatePicker from 'react-datepicker';
import { ThemeContext } from 'styled-components';
import { Button, ButtonStyle } from '../../atoms/Button';
import { Text, TextStyle } from '../../atoms/Text';
import {
  StyledActionsContainer,
  StyledCalendar,
  StyledCaption,
  StyledCaptionContainer,
  StyledContainer,
  StyledDatepicker,
  StyledInputIcon,
  StyledLabel,
  StyledWrapper
} from './Datepicker.styled';
import { Caption, DatepickerProps } from './Datepicker.types';

export const Datepicker = ({
  startDate,
  endDate,
  onChange,
  label,
  locale,
  minDate,
  maxDate,
  disabled,
  excludeDates,
  selectsRange,
  inline,
  readOnly,
  ...props
}: DatepickerProps) => {
  const theme = useContext(ThemeContext) || defaultTheme;

  const Container = ({ children }: { children: ReactNode }) => (
    <StyledContainer theme={theme}>
      <StyledCalendar theme={theme}>{children}</StyledCalendar>
    </StyledContainer>
  );

  // eslint-disable-next-line @typescript-eslint/no-shadow
  const renderDatepickerCaptions = (captions: Caption[]) => (
    <StyledCaptionContainer>
      {captions.map((caption, index) => (
        // eslint-disable-next-line react/no-array-index-key
        <StyledCaption color={caption.color} border={caption.border} key={`datepicker-caption-${index}`}>
          <Text textStyle={TextStyle.SmallDefault}>{caption.copy}</Text>
        </StyledCaption>
      ))}
    </StyledCaptionContainer>
  );

  const renderActions = () => (
    <StyledActionsContainer>
      {props.secondaryAction && <Button style={ButtonStyle.Secondary} text={props.secondaryAction.copy} fullWidth />}
      {props.primaryAction && <Button style={ButtonStyle.Primary} text={props.primaryAction.copy} fullWidth />}
    </StyledActionsContainer>
  );

  return (
    <TypographyProvider>
      <StyledWrapper theme={theme}>
        {!inline && (
          <StyledLabel disabled={disabled} htmlFor="datepicker-input" theme={theme}>
            {label}
          </StyledLabel>
        )}
        <StyledDatepicker disabled={disabled} readOnly={readOnly} theme={theme}>
          {!inline && (
            <StyledInputIcon disabled={disabled} readOnly={readOnly} theme={theme}>
              <Calendar size={IconSize.Medium} color={IconColor.DigitalBlack} />
            </StyledInputIcon>
          )}

          <DatePicker
            selected={startDate}
            startDate={startDate}
            endDate={selectsRange ? endDate : null}
            disabled={disabled}
            readOnly={readOnly}
            onChange={onChange}
            locale={locale}
            showPopperArrow={false}
            calendarClassName="calendar"
            calendarContainer={Container}
            disabledKeyboardNavigation
            calendarStartDay={1}
            useWeekdaysShort
            minDate={minDate}
            maxDate={maxDate}
            excludeDates={excludeDates}
            className="datepickerInput"
            inline={inline}
            selectsRange={selectsRange}
            shouldCloseOnSelect={!selectsRange}
            // eslint-disable-next-line react/jsx-props-no-spreading
            {...props}
          >
            {props.captions && renderDatepickerCaptions(props.captions)}
            {(props.primaryAction || props.secondaryAction) && renderActions()}
          </DatePicker>
        </StyledDatepicker>
      </StyledWrapper>
    </TypographyProvider>
  );
};
