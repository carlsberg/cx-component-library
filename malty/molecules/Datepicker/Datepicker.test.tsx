import { render } from '@carlsberg/malty.utils.test';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { Datepicker } from './Datepicker';
import { DatepickerProps } from './Datepicker.types';

const defaultProps: Required<Pick<DatepickerProps, 'label' | 'onChange' | 'startDate' | 'placeholderText'>> &
  DatepickerProps = {
  label: 'Select date',
  onChange: jest.fn(),
  startDate: new Date(2022, 0, 1),
  placeholderText: 'datepicker'
};

jest.mock('uuid', () => ({ v4: () => '00000000-0000-0000-0000-000000000000' }));

describe('datepicker', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render elements', () => {
    render(<Datepicker {...defaultProps} />);

    expect(screen.getByLabelText(defaultProps.label)).toBeInTheDocument();
    expect(screen.getByDisplayValue('01/01/2022')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('datepicker')).toBeInTheDocument();
  });

  it('should render datepicker with range', () => {
    render(<Datepicker {...defaultProps} selectsRange endDate={new Date(2022, 0, 3)} />);

    expect(screen.getByLabelText(defaultProps.label)).toBeInTheDocument();
    expect(screen.getByDisplayValue('01/01/2022 - 01/03/2022')).toBeInTheDocument();
  });

  it('should open datepicker when clicked', () => {
    render(<Datepicker {...defaultProps} />);

    expect(screen.queryByText('January')).not.toBeInTheDocument();

    userEvent.click(screen.getByLabelText(defaultProps.label));

    expect(screen.getByText('January')).toBeInTheDocument();
  });

  it('should render datepicker within a Portal and close it when selecting date', () => {
    render(<Datepicker {...defaultProps} withPortal />);

    userEvent.click(screen.getByLabelText(defaultProps.label));

    expect(screen.getByText('January')).toBeInTheDocument();

    const dayToBeClicked = screen.getByText('10');

    userEvent.click(dayToBeClicked);

    expect(defaultProps.onChange).toHaveBeenCalled();
    expect(screen.queryByText('January')).not.toBeInTheDocument();
  });

  it('should render datepicker within a Portal and not close it when selecting date if shouldCloseOnSelect is false', () => {
    render(<Datepicker {...defaultProps} withPortal shouldCloseOnSelect={false} />);
    userEvent.click(screen.getByLabelText(defaultProps.label));

    expect(screen.getByText('January')).toBeInTheDocument();

    const dayToBeClicked = screen.getByText('10');

    userEvent.click(dayToBeClicked);

    expect(defaultProps.onChange).toHaveBeenCalled();
    expect(screen.getByText('January')).toBeInTheDocument();
  });

  it('should not open datepicker when disabled', () => {
    render(<Datepicker {...defaultProps} disabled />);

    expect(screen.queryByText('January')).not.toBeInTheDocument();

    userEvent.click(screen.getByLabelText(defaultProps.label));

    expect(defaultProps.onChange).not.toHaveBeenCalled();
    expect(screen.queryByText('January')).not.toBeInTheDocument();
  });

  it('should not open datepicker when readOnly', () => {
    render(<Datepicker {...defaultProps} readOnly />);

    expect(screen.queryByText('January')).not.toBeInTheDocument();

    userEvent.click(screen.getByLabelText(defaultProps.label));

    expect(defaultProps.onChange).not.toHaveBeenCalled();

    expect(screen.queryByText('January')).not.toBeInTheDocument();
  });

  it('should not set a certain date when its off the minmax date interval', () => {
    render(
      <Datepicker
        {...defaultProps}
        startDate={new Date(2023, 8, 5)}
        minDate={new Date(2023, 8, 1)}
        maxDate={new Date(2023, 8, 20)}
      />
    );

    userEvent.click(screen.getByLabelText(defaultProps.label));

    const dayToBeClicked = screen.getByText('25');

    expect(screen.getByText('September')).toBeInTheDocument();

    userEvent.click(dayToBeClicked);

    expect(defaultProps.onChange).not.toHaveBeenCalled();
    expect(dayToBeClicked).toHaveAttribute('aria-disabled', 'true');
  });

  it('should be required to fill when required', () => {
    render(<Datepicker {...defaultProps} required />);

    expect(screen.getByLabelText(defaultProps.label)).toBeRequired();
  });

  it('should display the datepicker input according to the date format provided', () => {
    render(<Datepicker {...defaultProps} startDate={new Date(2022, 11, 1)} dateFormat="dd/MM/yyyy" />);

    expect(screen.getByDisplayValue('01/12/2022')).toBeInTheDocument();
  });

  it('should call `onClose` when datepicker is closed', () => {
    const onClose = jest.fn();

    render(<Datepicker {...defaultProps} startDate={new Date(2023, 12, 15)} onClose={onClose} />);

    userEvent.click(screen.getByLabelText(defaultProps.label));

    expect(onClose).not.toHaveBeenCalled();

    const dayToBeClicked = screen.getByText('16');

    userEvent.click(dayToBeClicked);

    expect(onClose).toHaveBeenCalledTimes(1);
  });

  it('should call `onClose` when datepicker is closed using selectsRange', () => {
    const onClose = jest.fn();

    render(<Datepicker {...defaultProps} startDate={new Date(2023, 12, 15)} onClose={onClose} selectsRange />);

    userEvent.click(screen.getByLabelText(defaultProps.label));

    expect(onClose).not.toHaveBeenCalled();

    const firstDay = screen.getByText('16');
    const secondDay = screen.getByText('17');

    userEvent.click(firstDay);
    userEvent.click(secondDay);

    expect(onClose).toHaveBeenCalledTimes(1);
  });

  it('should call `onClose` when clicking on the primary action', () => {
    const onClose = jest.fn();

    render(<Datepicker {...defaultProps} onClose={onClose} primaryAction={{ label: 'Confirm', action: jest.fn() }} />);

    userEvent.click(screen.getByLabelText(defaultProps.label));

    expect(onClose).not.toHaveBeenCalled();

    const confirmButton = screen.getByRole('button', { name: 'Confirm' });

    userEvent.click(confirmButton);

    expect(onClose).toHaveBeenCalledTimes(1);
  });

  it('should call `onClose` when clicking on the secondary action', () => {
    const onClose = jest.fn();

    render(<Datepicker {...defaultProps} onClose={onClose} secondaryAction={{ label: 'Cancel', action: jest.fn() }} />);

    userEvent.click(screen.getByLabelText(defaultProps.label));

    expect(onClose).not.toHaveBeenCalled();

    const cancelButton = screen.getByRole('button', { name: 'Cancel' });

    userEvent.click(cancelButton);

    expect(onClose).toHaveBeenCalledTimes(1);
  });
});
