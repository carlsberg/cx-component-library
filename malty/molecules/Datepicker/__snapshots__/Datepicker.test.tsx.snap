// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`datepicker matches snapshot 1`] = `
<span
  className="sc-bczRLJ hROJNc"
>
  <div
    className="sc-hKMtZM gZcLXw"
  >
    <label
      className="sc-dkzDqf gSBfxx"
      htmlFor="datepicker-input"
    >
      Select date
    </label>
    <div
      className="sc-eCYdqJ itCJjG"
    >
      <span
        className="sc-jSMfEi eSGvOq"
      >
        <svg
          className="sc-gsnTZi ivMBgY"
          color="digital-black"
          data-testid="svg-component"
          size="24px"
          version="1.1"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
          xmlnsXlink="http://www.w3.org/1999/xlink"
        >
          <g
            fillRule="evenodd"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M17 1a1 1 0 01.993.883L18 2v1h2a3 3 0 012.995 2.824L23 6v13a3 3 0 01-3 3H4a3 3 0 01-3-3V6a3 3 0 013-3h2V2a1 1 0 011.993-.117L8 2v1h8V2a1 1 0 011-1zm4 9H3v9a1 1 0 00.883.993L4 20h16a1 1 0 001-1zM9 16a1 1 0 010 2H7a1 1 0 010-2zm8 0a1 1 0 010 2h-2a1 1 0 010-2zm-8-4a1 1 0 010 2H7a1 1 0 010-2zm8 0a1 1 0 010 2h-2a1 1 0 010-2zM6 5H4a1 1 0 00-1 1v2h18V6a1 1 0 00-.883-.993L20 5h-2v1a1 1 0 01-1.993.117L16 6V5H8v1a1 1 0 01-1.993.117L6 6z"
            />
          </g>
        </svg>
      </span>
      <div
        className="react-datepicker-wrapper"
      >
        <div
          className="react-datepicker__input-container"
        >
          <input
            className="datepickerInput"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            readOnly={false}
            type="text"
            value="01/01/1970"
          />
        </div>
      </div>
    </div>
  </div>
</span>
`;
